### ISSUE TITLE
Typing a store via an interface by extending `State` silently breaks when using `getState()` syntax

### ISSUE THREAD

**Issue by ivancuric** [pmndrs/zustand#342]
Example:

Consider the following example, as described in the docs (https://github.com/pmndrs/zustand#typescript):
```tsx
type BearState = {
  bears: number;
  increase: (by: number) => void;
};

const useStore = create<BearState>((set) => ({
  bears: 0,
  increase: (by) => set((state) => ({ bears: state.bears + by }))
}));
```

Calling the following line will throw a typescript error:
```ts
useStore.getState().iDoNotExist;
// Property 'iDoNotExist' does not exist on type 'BearState'.ts(2339)
```

However if using an interface, it will silently fail:
```tsx
interface BearState extends State {
  bears: number;
  increase: (by: number) => void;
}
```
and `useStore.getState().iDoNotExist` will resolve to `unknown`

Demo:
https://codesandbox.io/s/zustand-interface-856gt?file=/src/App.tsx

**Comment by dai-shi** [comment]
Looks like a TS behavior?
Not sure if we can fix this without making `State` stricter.

https://www.typescriptlang.org/play?#code/C4TwDgpgBAysCGxoF4oCUIGMD2AnAJgDwDOwuAlgHYDmUAPlJQK4C2ARhLvVMSO9gBsANFCaUA1pWwB3SgD4AUAtCQoAIQjxccRCigBvBVCgctxAFyNWHXAG4jUKplybiESwAo2IS83acASihkOSgAN2xyfHsAX3sFHEpSHgQkSw0tHSRggxNNXAsABhEnF3g3TyCQgxioGITsJOAoaghgLL0PAPT8juDQj308s3Nix0pnV3coLv6auoClUl0AOnIAEWwAOWxgAFEAD3JSBVb21Igutc2d-aOThSokXAAzeExoDO0LgCYoCAOSEo+GIsAuBgcpgKvmsnHsxlKU083hh-lwVVCESisSUiWSyyQPx6mV+OSGUKKJQmZQqMwx83qeOaZw6f1QXWJ310bIG5PylPGk3K01m1X0tRiiwUBIgP2u212h2OwFObVZVw2CruyqUT04bw+6l6FwAzBDjBTUTZ4YKaSKUVY0fSsdEFPUGk0UroTZyOmbUHyRmNEcLKnNxXUPckWaachyjSTvXNBsNocHqUi6eGJVKZSb5bclScY96NTdFfcVUoFEA

**Comment by premshree** [comment]
@ivancuric since State is typed as

```ts
type State = Record<string | number | symbol, unknown>
```

– explicitly creating an interface that `extends State` is telling TypeScript that the type you want to constrain state to _includes_ any key satisfying the type `string|number|symbol`. Change `State` to the following:

```
type State = Record<number | symbol, number>
```
– and you'll notice that TS complains because you've changed the constraints you're extending.

When you pass a type to `create`, the TypeScript compiler is only making sure that the passed state _conforms_ to `State`:

```ts
export default function create<TState extends State>(
  createState: StateCreator<TState>
): StoreApi<TState> {
...
```

IOW, you shouldn't ever need to explicitly create an interface that extends `State` – what that is doing is the equivalent of:

```ts
type BearState = State & {
  bears: number;
  increase: (by: number) => void;
};
```

But perhaps there is a use case of extending `State` I might be missing?

**Comment by dai-shi** [comment]
> But perhaps there is a use case of extending State I might be missing?

Otherwise, it will complain elsewhere.

tbh, I prefer avoiding this loose `State`. (or can be said prefer `type State = unknown`)
this is more like a historical reason. I'm not sure how this stricter type change affects to lots of users. (so, i can't push forward.)

**Comment by hasparus** [comment]
I've just encountered a similar issue.

I can't use a state type defined with interface, becaue `State` has an index signature.

```
Type 'GameState' does not satisfy the constraint 'State'.
  Index signature is missing in type 'GameState'.ts(2344)
```

Instead od `extends State` zustand functions could use `extends object`.

**Comment by dai-shi** [comment]
https://codesandbox.io/s/zustand-interface-forked-e0xp4
Fixed in v3.5.0 by #376
